src/pybind/mgr/restful/module.py

```diff
@@ -0,0 +1,501 @@
+        pools = dict(map(lambda x: (x['pool'], x), self.get('osd_map')['pools']))
+        crush_rules = self.get('osd_map_crush')['rules']
+
+        osds_by_pool = {}
+        for pool_id, pool in pools.items():
+            pool_osds = None
+            for rule in [r for r in crush_rules if r['ruleset'] == pool['crush_ruleset']]:
+                if rule['min_size'] <= pool['size'] <= rule['max_size']:
+                    pool_osds = common.crush_rule_osds(self.get('osd_map_tree')['nodes'], rule)
+
+            osds_by_pool[pool_id] = pool_osds
+
+        for pool_id in pools.keys():
+            for in_pool_id in osds_by_pool[pool_id]:
+                osds[in_pool_id].append(pool_id)
+
+        return osds
+
+
+    def get_osds(self, pool_id=None, ids=None):
+        # Get data
+        osd_map = self.get('osd_map')
+        osd_metadata = self.get('osd_metadata')
+
+        # Update the data with the additional info from the osd map
+        osds = osd_map['osds']
+
+        # Filter by osd ids
+        if ids is not None:
+            osds = filter(
+                lambda x: str(x['osd']) in ids,
+                osds
+            )
+
+        # Get list of pools per osd node
+        pools_map = self.get_osd_pools()
+
+        # map osd IDs to reweight
+        reweight_map = dict([
+            (x.get('id'), x.get('reweight', None))
+            for x in self.get('osd_map_tree')['nodes']
+        ])
+
+        # Build OSD data objects
+        for osd in osds:
+            osd['pools'] = pools_map[osd['osd']]
+            osd['server'] = osd_metadata.get(str(osd['osd']), {}).get('hostname', None)
+
+            osd['reweight'] = reweight_map.get(osd['osd'], 0.0)
+
+            if osd['up']:
+                osd['valid_commands'] = common.OSD_IMPLEMENTED_COMMANDS
+            else:
+                osd['valid_commands'] = []
+
+        # Filter by pool
+        if pool_id:
+            pool_id = int(pool_id)
+            osds = filter(
+                lambda x: pool_id in x['pools'],
+                osds
+            )
+
+        return osds
+
+
+    def get_osd_by_id(self, osd_id):
+        osd = filter(
+            lambda x: x['osd'] == osd_id,
+            self.get('osd_map')['osds']
+        )
+
+        if len(osd) != 1:
+            return None
+
+        return osd[0]
+
+
+    def get_pool_by_id(self, pool_id):
+        pool = filter(
+            lambda x: x['pool'] == pool_id,
+            self.get('osd_map')['pools'],
+        )
+
+        if len(pool) != 1:
+            return None
+
+        return pool[0]
+
+
+    def submit_request(self, _request, **kwargs):
+        request = CommandsRequest(_request)
+        with self.requests_lock:
+            self.requests.append(request)
+        if kwargs.get('wait', 0):
+            while not request.is_finished():
+                time.sleep(0.001)
+        return request
+
+
+    def run_command(self, command):
```

Comment:
You are right. It was used by the cephx-based authentication to get info about the key right away but that is no longer the case so it is unused.